pkg/template/servicebroker/bind.go

```diff
@@ -1,123 +1,177 @@
 package servicebroker
 
 import (
+	"bytes"
+	"encoding/base64"
 	"errors"
+	"fmt"
 	"net/http"
-	"strconv"
+	"reflect"
 	"strings"
 
 	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/selection"
 	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/client-go/util/jsonpath"
 	kapi "k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/apis/authorization"
 
 	"github.com/golang/glog"
 	"github.com/openshift/origin/pkg/authorization/util"
 	"github.com/openshift/origin/pkg/openservicebroker/api"
+	routeapi "github.com/openshift/origin/pkg/route/api"
 	templateapi "github.com/openshift/origin/pkg/template/api"
 )
 
-// copied from vendor/k8s.io/kubernetes/pkg/kubelet/envvars/envvars.go
-// TODO: when the API for returning information from the bind call is cleaned up
-// and we're no longer temporarily using the environment variable style, remove
-// this.
-func makeEnvVariableName(str string) string {
-	return strings.ToUpper(strings.Replace(str, "-", "_", -1))
-}
-
-// getServices returns environment variable style details for all services
-// created by a given template in its namespace.  This API may not currently be
-// considered stable.  TODO: if a template creates a service in another
-// namespace, its details will not currently be returned.
-func (b *Broker) getServices(u user.Info, namespace, instanceID string) (map[string]string, *api.Response) {
-	glog.V(4).Infof("Template service broker: getServices")
-
-	requirement, _ := labels.NewRequirement(templateapi.TemplateInstanceLabel, selection.Equals, []string{instanceID})
+func evaluateJSONPathExpression(obj interface{}, annotation, expression string, base64encode bool) (string, error) {
+	var s []string
 
-	if err := util.Authorize(b.kc.Authorization().SubjectAccessReviews(), u, &authorization.ResourceAttributes{
-		Namespace: namespace,
-		Verb:      "list",
-		Group:     kapi.GroupName,
-		Resource:  "services",
-	}); err != nil {
-		return nil, api.Forbidden(err)
+	j := jsonpath.New("templateservicebroker")
+	err := j.Parse(expression)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse annotation %s: %v", annotation, err)
 	}
 
-	serviceList, err := b.kc.Core().Services(namespace).List(metav1.ListOptions{LabelSelector: labels.NewSelector().Add(*requirement).String()})
+	results, err := j.FindResults(obj)
 	if err != nil {
-		if kerrors.IsForbidden(err) {
-			return nil, api.Forbidden(err)
+		return "", fmt.Errorf("FindResults failed on annotation %s: %v", annotation, err)
+	}
+
+	for _, r := range results {
+		switch len(r) {
+		case 0:
+			// shouldn't happen
```

Comment:
do we not want to at least log an error indicating what the jsonpath was and the object we searched, if this happens then?